readdir.go

```diff
@@ -0,0 +1,58 @@
+package hdfs
+
+import (
+	"context"
+	"github.com/beyondstorage/go-storage/v4/types"
+	"os"
+)
+
+type listDirInput struct {
+	rp  string
+	dir string
+
+	fileList []os.FileInfo
+
+	continuationToken string
+	counter           int
+}
+
+func (i *listDirInput) ContinuationToken() string {
+	return i.continuationToken
+}
+
+func (s *Storage) listDirNext(ctx context.Context, page *types.ObjectPage) (err error) {
+	input := page.Status.(*listDirInput)
+
+	input.fileList, err = s.hdfs.ReadDir(input.rp)
```

Comment:
`s.hdfs.ReadDir(input.rp)` already read all content at once, see code here: https://github.com/colinmarc/hdfs/blob/master/readdir.go

```go
func (c *Client) ReadDir(dirname string) ([]os.FileInfo, error) {
	f, err := c.Open(dirname)
	if err != nil {
		return nil, err
	}

	return f.Readdir(0)
}
```

Instead, we need to open the dir and read them via count.

And we don't need to store fileList/counter in `listDirInput`.

Reply:
So count is up to us? For example `const defaultListObjectCount=100`

Reply:
What we need is `limit` instead of `count`, there is no need for `list` to calculate the count.

We only need to `f.Readdir(limit)` and return `IteratorDone` if we meet `io.EOF`.

Ref: https://github.com/colinmarc/hdfs/blob/master/file_reader.go#L255-L257

Reply:
Yes, this is slightly confusing, and I thought of using a constant to represent the read upper limit? Maybe there is something wrong with my description